python/lsst/daf/butler/_butler.py

```diff
@@ -1463,9 +1463,13 @@ def ingest(self, *datasets: FileDataset, transfer: Optional[str] = "auto", run:
         allResolvedRefs: List[DatasetRef] = []
         for datasetType, groupForType in progress.iter_item_chunks(groupedData.items(),
                                                                    desc="Bulk-inserting datasets by type"):
-            refs = self.registry.insertDatasets(datasetType,
-                                                dataIds=groupForType.keys(),
-                                                run=run)
+            refs = self.registry.insertDatasets(
+                datasetType,
+                dataIds=groupForType.keys(),
+                run=run,
+                # TODO: ask Datastore if it needs expansion instead.
+                expand=(transfer not in ("direct", None)),
```

Comment:
It would possibly be better to ask datastore the question as to whether it needs expanded dataIds or not given the transfer mode, rather than assuming in butler level but I'm sure we can sort this out later.

Reply:
How about I add this to base `Datastore` and use it here and in `obs_base`:
```
@abstractmethod
def needs_expanded_data_ids(
    self,
    transfer: Optional[str],
    entity: Optional[Union[DatasetRef, DatasetType, StorageClass]] = None,
) -> bool:
    """Test whether this datastore needs expanded data IDs to ingest the given entity.

    Parameters
    ----------
    transfer : `str` or `None`
        Transfer mode for ingest.
    entity, optional
        Object representing what will be ingested.  If not provided (or less informative),
        `True` may be returned even if expanded data IDs aren't necessary.

    Returns
    -------
    needed : `bool`
        If `True`, expanded data IDs may be needed.  `False` only if expansion
        definitely isn't necessary.
    """
    raise NotImplementedError()
```

and use `(transfer not in ("direct", None))` as the initial implementation in `FileDatastore`?

Reply:
That would work. I like you you added in a representative `entity` just in case we need to look up in a datastore config file.

Reply:
Done I also added implementations for `InMemoryDatastore` (`return False`) and `ChainedDatastore`, and the latter is worth a look - I don't think it's actually able to forward that `entity` argument on to the child datastores while still responding conservatively, which is a bit unfortunate but (I think) not a big deal.

Reply:
The chained implementation is fine I think.